### PR TITLE
feature/app-notification

### PR DESCRIPTION
**Pull request by AminIdeaUsher**
### Optimized App Notification Module
Creates only 1 notification per user group

To check if user has seen a notification:
`isSeen = user.appNotificationsLastSeenAt >= notification.scheduledAt`

To get status of a notification (sent / scheduled)
`isSent = notification.scheduledAt > currrentDateTime`

### CODE REVIEW

File: src/validations/appNotification.validation.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  query: Joi.object().keys({
+    ...dbOptionsSchema,
+  }),
+};
+
+const getAppNotificationsSentByAdmin = getAppNotifications;
+
+const deleteAppNotification = {
+  params: Joi.object().keys({
+    appNotificationId: Joi.string()
```
Comment (Tushar-4781) [inline src/validations/appNotification.validation.js]: just name it id

File: src/routes/v1/appNotification.route.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  validate(appNotificationValidation.getAppNotificationsSentByAdmin),
+  appNotificationController.getAppNotificationsSentByAdmin
+);
+
+// Update notifications last seen at for current user
+router.patch('/seen', firebaseAuth('All'), appNotificationController.updateAppNotificationsLastSeenAt);
+
+// Delete a notification
+router.delete(
+  '/:appNotificationId',
```
Comment (Tushar-4781) [inline src/routes/v1/appNotification.route.js]: you can name it id its understood by the standards that notifications/:id -> is a document inside notifications. simplify the variable naming

File: src/controllers/appNotification.controller.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  const appNotification = await appNotificationService.createAppNotification({
+    ...req.body,
+    isCreatedByAdmin: true,
+  });
+  res.status(httpStatus.CREATED).json({data: appNotification});
+});
+
+const getAppNotifications = catchAsync(async (req, res) => {
+  const {options} = getPaginateConfig(req.query);
+  options.sortBy = 'createdAt';
```
Comment (Tushar-4781) [inline src/controllers/appNotification.controller.js]: you can set these as default in validation schema

File: src/controllers/appNotification.controller.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  options.sortOrder = 'desc';
+
+  const appNotifications = await appNotificationService.getAppNotifications(
+    {
+      $or: [{targetRole: {$in: [userTypes.ALL, req.user.__t]}}, {user: req.user._id}],
```
Comment (Tushar-4781) [inline src/controllers/appNotification.controller.js]: can we move it in service ?

File: src/controllers/appNotification.controller.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      scheduledAt: {$gte: req.user.createdAt},
+    },
+    options
+  );
+  res.json({data: appNotifications});
+});
+
+const getAppNotificationsSentByAdmin = catchAsync(async (req, res) => {
+  const {filters, options} = getPaginateConfig(req.query);
+  const appNotifications = await appNotificationService.getAppNotifications(
```
Comment (Tushar-4781) [inline src/controllers/appNotification.controller.js]: name the final variable "data" that has to be send. less code better readability

**Review by Tushar-4781 — CHANGES_REQUESTED** [review]

File: src/controllers/appNotification.controller.js (hunk 3 above)
Comment (AminIdeaUsher) [reply src/controllers/appNotification.controller.js]: Just saw that its already the default

File: src/services/appNotification.service.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  AppNotification.paginate(
+    {
+      ...filters,
+      $or: [{targetRole: {$in: [userTypes.ALL, user.__t]}}, {user: user._id}],
+      scheduledAt: {$gte: user.createdAt},
+    },
+    options
+  );
+
+const deleteAppNotification = id => AppNotification.findByIdAndDelete(id);
```
Comment (Tushar-4781) [inline src/services/appNotification.service.js]: deleteAppNotificationById

**Review by Tushar-4781 — CHANGES_REQUESTED** [review]

**Review by Tushar-4781 — APPROVED** [review]

**Pull request merged**